test/langtools/tools/javac/bang/BangTypesCompilationTests.java

```diff
@@ -318,7 +465,114 @@ void test(List<? super String!> ls2) {
                                 }
                                 """,
                                 Result.Clean,
-                                "" /* no warnings in this case */)
+                                ""),
+                        new DiagAndCode(
+                                """
+                                interface Shape {}
+                                value class Point implements Shape {}
+                                class Box<T> {}
+                                class Test {
+                                    void m(Box<Point!> lp) {
+                                        foo(lp);
+                                    }
+                                    void foo(Box<? extends Shape> ls) {}
+                                }
+                                """,
+                                Result.Clean,
+                                ""),
+                        new DiagAndCode(
+                                """
+                                interface Shape {}
+                                value class Point implements Shape {}
+                                class Box<T> {}
+                                class Test {
+                                    void m(Box<Shape!> lp) {
```

Comment:
This test reminds me of something: capture conversion should probably be tweaked to take nullability into account. Consider this:

```
class Foo<X extends Bar!> { 
    X x() { ... }
}

Foo<? extends Object> foo = ...
Object! restricted = foo.x(); // this should be ok  (e.g. no warnings)
```

I'm basing this on the fact that capture conversion already tries to establish some common facts between declaration-site and use-site. This test is similar to when use-site says `<? extends Object>` but the decl site says `<X extends List>`, in which case `List` is used as the bound for the captured variable. In this case, it seems like the null restriction that is present at the declaration site should be carried over to the captured variable as well.  Of course this is all outside the context of this review, but I thought I'd leave a comment to note the issue.

Reply:
yes, interesting